examples/add_points_on_nD_shapes.py

```diff
@@ -0,0 +1,60 @@
+"""Add points on nD shapes in 3D using a mouse callback"""
+import napari
+import numpy as np
+
+# Create rectangles in 4D
+shapes_data = np.array(
+    [
+        [
+            [0, 50, 75, 75],
+            [0, 50, 125, 75],
+            [0, 100, 125, 125],
+            [0, 100, 75, 125]
+        ],
+        [
+            [0, 10, 75, 75],
+            [0, 10, 125, 75],
+            [0, 40, 125, 125],
+            [0, 40, 75, 125]
+        ],
+        [
+            [1, 100, 75, 75],
+            [1, 100, 125, 75],
+            [1, 50, 125, 125],
+            [1, 50, 75, 125]
+        ]
+    ]
+)
+
+# add an empty 4d points layer
+viewer = napari.view_points(ndim=4, size=3)
+points_layer = viewer.layers[0]
+
+# add the shapes layer to the viewer
+properties = {'index': [0, 1, 2]}
+shapes_layer = viewer.add_shapes(
+    shapes_data,
+    face_color=['magenta', 'green', 'blue'],
+    edge_color='white',
+    blending='additive',
+    properties=properties,
+    text='index'
+)
+
+
+@shapes_layer.mouse_drag_callbacks.append
+def on_click(layer, event):
+
+    shape_index, intersection_point = layer.get_index_and_intersection(
+        event.position,
+        event.view_direction,
+        event.dims_displayed
+    )
+
+    if (shape_index is not None) and (intersection_point is not None):
+        points_layer.add(intersection_point)
```

Comment:
I understand the motivation for this refactor but really do think this API is messier, it forces the user to be aware of
- the structure of the event object
- the dims model

I think if we want to keep this knowledge of the events off the layer (which I understand and support) then a better approach is to provide all interactivity utilities in an `interactivity_utils.py` file, have those easily accessible and always point to those

will not hold up merge for this, we can get this and click and drag in then refactor when all of the pieces are together

Reply:
ok yeah, we can continue to iterate here in future PRs